gateway/src/lib.rs

```diff
@@ -683,43 +676,150 @@ impl Actor {
+                // TODO: are sharing the nonce across the network, this might be slow
+                // TODO: as owners needs to execute one after anotehr and each owner
+                // TODO: might have a different speed and it might cause congestion?
+                if st.applied_topdown_nonce != cross_msg.msg.nonce {
+                    return Err(actor_error!(
+                        illegal_state,
+                        "the top-down message being applied doesn't hold the subsequent nonce"
+                    ));
+                }
 
-                // if directed to the current network propagate the message
                 if sto == st.network_name {
-                    // FIXME: Should we handle return in some way?
-                    let _ = cross_msg.send(rt, rto)?;
+                    rt.transaction(|st: &mut State, _| {
+                        st.applied_topdown_nonce += 1;
+                        Ok(())
+                    })?;
+
+                    // We can return the send result
+                    return cross_msg.send(rt, rto);
                 }
             }
             _ => {
                 return Err(actor_error!(
                     illegal_argument,
                     "cross-message to apply dosen't have the right type"
-                ));
+                ))
             }
         };
 
-        Ok(())
+        rt.transaction(|st: &mut State, rt| {
+            let owner = cross_msg
+                .msg
+                .from
+                .raw_addr()
+                .map_err(|_| actor_error!(illegal_argument, "invalid address"))?;
+            st.insert_postbox(
+                rt.store(),
+                Some(vec![owner]),
+                MIN_CROSS_MSG_GAS.clone(),
+                cross_msg,
+            )
+            .map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "error save topdown messages")
+            })?;
+            Ok(())
+        })?;
+
+        Ok(RawBytes::default())
+    }
+
+    fn propagate<BS, RT>(rt: &mut RT, params: PropagateParams) -> Result<RawBytes, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        // TODO: update this to EOA
+        rt.validate_immediate_caller_accept_any()?;
+
+        let PropagateParams { gas, postbox_cid } = params;
+        let owner = rt.message().caller();
+
+        rt.transaction(|st: &mut State, rt| {
+            let postbox_item = st.load_from_postbox(rt.store(), postbox_cid).map_err(|e| {
+                log::error!("encountered error loading from postbox: {:?}", e);
+                actor_error!(unhandled_message, "cannot load from postbox")
+            })?;
+
+            if let Some(owners) = postbox_item.owners && !owners.contains(&owner) {
+                return Err(actor_error!(illegal_state, "owner not match"));
+            }
+
+            if postbox_item.gas < gas {
```

Comment:
No need to perform this check, we should remove the `gas` field from cross-message items. This is determined by the actor.